targets/TARGET_Realtek/TARGET_AMEBA/serial_api.c

```diff
@@ -0,0 +1,872 @@
+
+/******************************************************************************
+ * READ/WRITE
+ ******************************************************************************/
+
+int serial_getc(serial_t *obj) 
+{
+#ifdef CONFIG_MBED_ENABLED
+	if(obj->index == UART_3){
+		return log_uart_getc(&stdio_uart_log);
+	}
+#endif
+    PHAL_RUART_ADAPTER pHalRuartAdapter=(PHAL_RUART_ADAPTER)&(obj->hal_uart_adp);
+    u8  uart_idx = pHalRuartAdapter->UartIndex;
+
+    while (!serial_readable(obj));
+    return (int)((HAL_RUART_READ32(uart_idx, RUART_REV_BUF_REG_OFF)) & 0xFF);
+}
+
+void serial_putc(serial_t *obj, int c) 
+{
+#ifdef CONFIG_MBED_ENABLED
+	if(obj->index == UART_3){
+		log_uart_putc(&stdio_uart_log, (char)c);
+		return;
+	}
+#endif
+    PHAL_RUART_ADAPTER pHalRuartAdapter=(PHAL_RUART_ADAPTER)&(obj->hal_uart_adp);
+    u8  uart_idx = pHalRuartAdapter->UartIndex;
+    
+    while (!serial_writable(obj));
+    HAL_RUART_WRITE32(uart_idx, RUART_TRAN_HOLD_REG_OFF, (c & 0xFF));
+
+    if (serial_irq_en[uart_idx] & SERIAL_TX_IRQ_EN) {
+        // UnMask TX FIFO empty IRQ
+        pHalRuartAdapter->Interrupts |= RUART_IER_ETBEI;
+        HalRuartSetIMRRtl8195a (pHalRuartAdapter);
+    }
+}
+
+int serial_readable(serial_t *obj) 
+{
+#ifdef CONFIG_MBED_ENABLED
+	if(obj->index == UART_3){
+		return log_uart_readable(&stdio_uart_log);
+	}
+#endif
+
+    PHAL_RUART_ADAPTER pHalRuartAdapter=(PHAL_RUART_ADAPTER)&(obj->hal_uart_adp);
+    u8  uart_idx = pHalRuartAdapter->UartIndex;
+
+    if ((HAL_RUART_READ32(uart_idx, RUART_LINE_STATUS_REG_OFF)) & RUART_LINE_STATUS_REG_DR) {
+        return 1;
+    }
+    else {
+        return 0;
+    }
+}
+
+int serial_writable(serial_t *obj) 
+{
+#ifdef CONFIG_MBED_ENABLED
+	if(obj->index == UART_3){
+		return log_uart_writable(&stdio_uart_log);
+	}
+#endif
+
+    PHAL_RUART_ADAPTER pHalRuartAdapter=(PHAL_RUART_ADAPTER)&(obj->hal_uart_adp);
+    u8  uart_idx = pHalRuartAdapter->UartIndex;
+
+    if (HAL_RUART_READ32(uart_idx, RUART_LINE_STATUS_REG_OFF) & 
+                        (RUART_LINE_STATUS_REG_THRE)) {
+       return 1;
+    }
+    else {
+        return 0;
+    }
+}
+
+void serial_clear(serial_t *obj) 
+{
+#ifdef CONFIG_MBED_ENABLED
+	if(obj->index == UART_3){
+		log_uart_clear(&stdio_uart_log);
+		return;
+	}
+#endif
+
+    PHAL_RUART_ADAPTER pHalRuartAdapter;
+
+    pHalRuartAdapter = &(obj->hal_uart_adp);
+    HalRuartResetTRxFifo((VOID *)pHalRuartAdapter);
+}
+
+#ifndef CONFIG_MBED_ENABLED
+void serial_clear_tx(serial_t *obj) 
+{
+    PHAL_RUART_ADAPTER pHalRuartAdapter;
+
+    pHalRuartAdapter = &(obj->hal_uart_adp);
+    HalRuartResetTxFifo((VOID *)pHalRuartAdapter);
+}
+
+void serial_clear_rx(serial_t *obj) 
+{
+    PHAL_RUART_ADAPTER pHalRuartAdapter;
+
+    pHalRuartAdapter = &(obj->hal_uart_adp);
+    HalRuartResetRxFifo((VOID *)pHalRuartAdapter);
+}
+#endif
+
+void serial_break_set(serial_t *obj) 
+{
+#ifdef CONFIG_MBED_ENABLED
+	if(obj->index == UART_3){
+		log_uart_break_set(&stdio_uart_log);
```

Comment:
please use 4 spaces as in the rest of the code, there are few lines here using 8 as I can see